main.go

```diff
@@ -279,18 +280,24 @@ func main() {
 
 	//+kubebuilder:scaffold:builder
 
-	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
+	if err = mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
 		setupLog.Error(err, "unable to set up health check")
 		os.Exit(1)
 	}
-	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
+	if err = mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
 		setupLog.Error(err, "unable to set up ready check")
 		os.Exit(1)
 	}
 
-	setupLog.Info("starting manager")
-	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
-		setupLog.Error(err, "problem running manager")
+	client, err := dynamic.NewForConfig(mgr.GetConfig())
```

Comment:
I don't think there's anything missing in controller-runtime's. Rather, it's more about what it adds.

Right now, given how the Policy Machinery was designed, it doesn't add a lot (though also not nothing.) Generally the same operations are available; the dynamic client is also a cache client for read operations; and the structuring of objects is already built-in and thought for all read operations post-construction of the topology to first target the topology object.

IOW, no hard reason to depend on controller-runtime here IMO.

Maybe a couple diffs worth mentioning are:
| controller-runtime | kube/client-go |
|-|-|
| read/write separation | single client, though with reading operations discouraged as a pattern |
| implicit restructuring of resources from the concrete the Go types | uses declared [resources](https://github.com/Kuadrant/policy-machinery/blob/main/controller/resources.go) and explicit un/marshalling of `Unstructured`, though already handled built-in for the topology and [helper](https://github.com/Kuadrant/policy-machinery/blob/c4c99ba6f9e97306b4c3c1f3fcbb2825f561305d/controller/runnable.go#L213) provided for write operations |

Reply:
The code using this client seems a lot more confusing and cumbersome. Just a personal opinion. Additionally no need for a explicit un/marshalling. 

What do others think about this? cc @Boomatang @KevFan

Reply:
To be clear, I'm not against using controller-runtime's client, but if we do (off the top of my head):
1. The change to expect a controller-runtime client associated with the controller (instead of client-go's <a href="https://pkg.go.dev/k8s.io/client-
go@v0.30.2/dynamic#DynamicClient"><code>dynamic.DynamicClient</code></a>) needs to be done in the Policy Machinery first.

2. We cannot use [`mgr.GetClient()`](https://github.com/Kuadrant/kuadrant-operator/blob/b7a58cab263e38b105fcde5dc9670bfe73a73e05/main.go#L146C3-L146C18) as this returns a controller-runtime's [`cluster.Client`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.3/pkg/cluster#Cluster.GetClient). Instead, we'd need a new client created using [`NewWithWatch`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client#NewWithWatch).

3. In the Policy Machinery, we will no longer be able to instantiate runnables (informers/reconcilers) only from the singular type; we'd need to pass along in all the calls to [`controller.Watch[T Object](obj T, …)`](https://github.com/Kuadrant/policy-machinery/blob/c4c99ba6f9e97306b4c3c1f3fcbb2825f561305d/controller/runnable.go#L61) also an instance of a List type – i.e. sth like `controller.Watch[O Object, L ctrlruntimeclient.ObjectList](obj O, listObj L, …)`. This means a fairly heavy refactoring of https://github.com/Kuadrant/policy-machinery/blob/main/controller/runnable.go

Reply:
I am happy to do that heavy refactor in the policy machinery lib if people do not like doing

```go
cm := &corev1.ConfigMap{ ... }                                                                                                                          
                                                                                                                           
configMapRes := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "configmaps"}                              
unstructuredCM, _ := controller.Destruct(cm)                                                                               
 _, err := r.Client.Resource(configMapRes).Namespace(cm.Namespace).Create(ctx, unstructuredCM, metav1.CreateOptions{})  
```

I am here to serve people command

Reply:
I do think using the `DynamicClient` is cumbersome to a point but I also know this is the first time I have had to use it so it is also different. I don't know know if the lack of practice with the client leads to it feeling cumbersome. We always know there was going to be a learning curve with doing these changes so I am happy to let it cook. 

There is all my natural dislike of the controller-runtime. I believe it is a dependency that we can live without and using the `DynamicClient` reduces that dependency which can only be a good thing.